File: Day19/Part6/n_queens_ii.py
# LC 52 — N-Queens II
# Backtracking with bitmasks

def totalNQueens(n: int) -> int:
    maskAll = (1 << n) - 1
    ans = 0
    def dfs(row: int, cols: int, diag: int, anti: int):
        nonlocal ans
        if row == n:
            ans += 1
            return
        avail = maskAll & ~(cols | diag | anti)
        while avail:
            p = avail & -avail
            avail -= p
            dfs(row+1, cols | p, ((diag | p) << 1) & maskAll, ((anti | p) >> 1))
    dfs(0, 0, 0, 0)
    return ans

if __name__ == "__main__":
    print(totalNQueens(4))  # 2

File: Day19/Part6/n_queens_ii.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 52 — N-Queens II
// Backtracking with bitmasks

int n;
int maskAll;

int dfs(int row, int cols, int diag, int anti){
    if (row == n) return 1;
    int count = 0;
    int avail = maskAll & ~(cols | diag | anti);
    while(avail){
        int p = avail & -avail;
        avail -= p;
        count += dfs(row+1, cols | p, ((diag | p) << 1) & maskAll, ((anti | p) >> 1));
    }
    return count;
}

int totalNQueens(int _n){
    n = _n;
    maskAll = (1<<n) - 1;
    return dfs(0, 0, 0, 0);
}

int main(){
    cout << totalNQueens(4) << "\n"; // 2
    return 0;
}
